docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,364 @@
+
+```json
+{
+  "commands": ["kustomize edit set image my-app:v0.0.2", "kustomize build ."],
+  "drySHA": "ab2382f",
+  "commitAuthor": "Michael Crenshaw <michael@example.com>",
+  "commitMessage": "chore: bump Helm dependency chart to 32.1.12",
+  "repoURL": "https://github.com/argoproj/argocd-example-apps"
+}
+```
+
+To request a commit to the hydrated branch, the application controller will make a gRPC call to the CommitManifests service.
+
+A single call will bundle all the changes destined for a given targetBranch.
+
+It's the application controller's job to ensure that the user has write access to the repo before making the call.
+
+```protobuf
+// CommitManifests represents the caller's request for some Kubernetes manifests to be pushed to a git repository.
+message CommitManifests {
+  // repoURL is the URL of the repo we're pushing to. HTTPS or SSH URLs are acceptable.
+  required string repoURL = 1;
+  // targetBranch is the name of the branch we're pushing to.
+  required string targetBranch = 2;
+  // drySHA is the full SHA256 hash of the "dry commit" from which the manifests were hydrated.
+  required string drySHA = 3;
+  // commitAuthor is the name of the author of the dry commit.
+  required string commitAuthor = 4;
+  // commitMessage is the short commit message from the dry commit.
+  required string commitMessage = 5;
+  // commitTime is the dry commit timestamp.
+  required string commitTime = 6;
+  // details holds the information about the actual hydrated manifests.
+  repeated CommitPathDetails details = 7;
+}
+
+// CommitManifestDetails represents the details about a 
+message CommitPathDetails {
+  // path is the path to the directory to which these manifests should be written.
+  required string path = 1;
+  // manifests is a list of JSON documents representing the Kubernetes manifests.
+  repeated string manifests = 2;
+  // readme is a string which will be written to a README.md alongside the manifest.yaml. 
+  required string readme = 3;
+}
+
+message CommitManifestsResponse {
+}
+```
+
+### Use cases
+
+#### Use case 1:
+
+An organization with strong requirements around change auditing might enable manifest hydration in order to generate a full history of changes.
+
+#### Use case 2:
+
+### Implementation Details/Notes/Constraints
+
+### Detailed examples
+
+### Security Considerations
+
+This proposal would involve introducing a component capable of pushing to git.
+
+We'll need to consider what git permissions setup to recommend, what security features we should recommend enabling (e.g. branch protection), etc.
+
+We'll also need to consider how to store the git push secrets. It's probable that they'll need to be stored in a namespace separate from the other Argo CD components to provide a bit extra protection.
+
+### Risks and Mitigations
+
+### Upgrade / Downgrade Strategy
+
+## Drawbacks
+
+## Alternatives
```

Comment:
I think the multi-sources support of the hydrator should be described in depth as an alternative.

If the implementation does not plan on supporting multi-sources Application, the different use cases of multi-sources should be described with a possible implementation of the single Source hydrator and a path forward.

Otherwise, we should validate that the current hydrator spec is forward compatible with a future support.

Reply:
Added some more content about the multi-source question.